### PR TITLE
CRM-20855: Disabling "Search Primary Details Only" causes partial CiviMail delivery failure

### PR DESCRIPTION
**Pull request by monishdeb**
* [CRM-20855: Disabling "Search Primary Details Only" causes partial CiviMail delivery failure](https://issues.civicrm.org/jira/browse/CRM-20855)

### CODE REVIEW

**Comment by mfb** [comment]
Confirmed that this resolves the bug :+1:

**Comment by monishdeb** [comment]
Thanks @mfb and @lcdservices for your feedback 

@colemanw / @eileenmcnaughton can you merge this PR?

File: CRM/Contact/BAO/Query.php [hunk 1]
```diff
@@ -436,6 +442,11 @@ public function __construct(
       $this->_params = array();
     }
 
+    // Search by Primary location type ONLY if searchPrimaryDetailsOnly setting is on OR
+    //  search query is trigerred via api OR
+    //  special parameter search_by_primary_details_only = TRUE
+    self::$_alwaysSearchByPrimaryOnly = (Civi::settings()->get('searchPrimaryDetailsOnly') || $apiEntity || !empty($params['search_by_primary_details_only']));
```
Comment (eileenmcnaughton) [inline CRM/Contact/BAO/Query.php]: I think the api should set $params['search_by_primary_details_only'] rather than $apiEntity being code for that.

Comment (monishdeb) [inline CRM/Contact/BAO/Query.php]: @eileenmcnaughton didn't found the exact spot set ```search_by_primary_details_only = 1``` via all API params that uses location type. Reason why the earlier code used $apiEntity. Or can you suggest me we could set this parameter for API?

**Comment by eileenmcnaughton** [comment]
@monishdeb test is good & underlying logic is good. I'm less keen on what variables are being set where & what they are being relied on to do.

I put my thoughts in a separate PR https://github.com/civicrm/civicrm-core/pull/10915/files (wanted to see what will happen with tests on it) - but basically I think we should consolidate on existing parameters & be explicit about passing in searchPrimary rather than overloading the 'apiEntity' which already has an ambiguous relationship to 'mode'

**Comment by monishdeb** [comment]
Closing in favor of https://github.com/civicrm/civicrm-core/pull/10915

**Pull request closed without merging**